File: go/templates/sa_template.go
// go/templates/sa_template.go
package main

import (
	"fmt"
	"math"
	"math/rand"
	"os"
	"time"
)

const (
	T0    = 100.0
	Alpha = 0.995
)

func logScore(step, score int) {
	fmt.Fprintf(os.Stderr, "%d %d\n", step, score)
}

func sa(
	initial int,
	neighbor func(x int, rng *rand.Rand) (int, bool),
	score func(x int) int,
	limit time.Duration,
	rng *rand.Rand,
) (bestState, bestScore int) {
	start := time.Now()
	state := initial
	bestState = state
	bestScore = score(state)
	curScore := bestScore

	T := T0
	step := 0

	for time.Since(start) < limit {
		newState, ok := neighbor(state, rng)
		if !ok {
			continue
		}
		newScore := score(newState)
		delta := float64(newScore - curScore)

		if delta >= 0 || rng.Float64() < math.Exp(delta/T) {
			state = newState
			curScore = newScore
			if curScore > bestScore {
				bestScore = curScore
				bestState = state
			}
		}

		T *= Alpha
		step++
		if step%1000 == 0 {
			logScore(step, curScore)
		}
	}

	return bestState, bestScore
}

func main() {
	const target = 42
	rng := rand.New(rand.NewSource(42))

	score := func(x int) int {
		d := x - target
		return -(d * d)
	}

	neighbor := func(x int, rng *rand.Rand) (int, bool) {
		dx := 1
		if rng.Intn(2) == 0 {
			dx = -1
		}
		nxt := x + dx
		if nxt < 0 || nxt > 100 {
			return 0, false
		}
		return nxt, true
	}

	bestState, bestScore := sa(0, neighbor, score, 500*time.Millisecond, rng)
	fmt.Printf("best_state=%d best_score=%d\n", bestState, bestScore)
}

File: python/templates/sa_template.py
# python/templates/sa_template.py
import math
import random
import sys
import time
from typing import Callable, Optional, Tuple, TypeVar

T = TypeVar("T")

T0 = 100.0
ALPHA = 0.995


def log_score(step: int, score: int) -> None:
    print(step, score, file=sys.stderr)


def sa(
    initial_state: T,
    neighbor_fn: Callable[[T], Optional[T]],
    score_fn: Callable[[T], int],
    time_limit_sec: float = 2.0,
    rng: Optional[random.Random] = None,
) -> Tuple[T, int]:
    """neighbor_fn(state) -> new_state or None
    score_fn(state) -> int (high is better)
    """
    if rng is None:
        rng = random.Random()

    start = time.perf_counter()

    state = initial_state
    best_state = state
    best_score = score = score_fn(state)

    T = T0
    step = 0

    while time.perf_counter() - start < time_limit_sec:
        new_state = neighbor_fn(state)
        if new_state is None:
            continue
        new_score = score_fn(new_state)
        delta = new_score - score

        if delta >= 0 or rng.random() < math.exp(delta / T):
            state = new_state
            score = new_score
            if score > best_score:
                best_score = score
                best_state = state

        T *= ALPHA
        step += 1
        if step % 1000 == 0:
            log_score(step, score)

    return best_state, best_score


def demo() -> None:
    target = 42
    state = 0

    def score_fn(x: int) -> int:
        return -(x - target) ** 2

    def neighbor_fn(x: int) -> Optional[int]:
        dx = random.choice([-1, 1])
        nxt = x + dx
        if 0 <= nxt <= 100:
            return nxt
        return None

    best_state, best_score = sa(state, neighbor_fn, score_fn, time_limit_sec=0.5, rng=random.Random(42))
    print(f"best_state={best_state} best_score={best_score}")


if __name__ == "__main__":
    demo()
